lint/commentedOutCode_checker.go

```diff
@@ -0,0 +1,101 @@
+package lint
+
+//! Detects commented-out code inside function bodies.
+//
+// @Before:
+// // fmt.Println("Debugging hard")
+// foo(1, 2)
+//
+// @After:
+// foo(1, 2)
+
+import (
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"github.com/go-toolsmith/strparse"
+)
+
+func init() {
+	addChecker(&commentedOutCodeChecker{}, attrExperimental)
+}
+
+type commentedOutCodeChecker struct {
+	checkerBase
+}
+
+func (c *commentedOutCodeChecker) VisitLocalComment(cg *ast.CommentGroup) {
+	s := cg.Text() // Collect text once
+
+	// We do multiple heuristics to avoid false positives.
+	// Many things can be improved here.
+
+	markers := []string{
+		"TODO", // TODO comments with code are permitted.
+
+		// "http://" is interpreted as a label with comment.
+		// There are other protocols we might want to include.
+		"http://",
+		"https://",
+
+		"e.g. ", // Clearly not a "selector expr" (mostly due to extra space)
+	}
+	for _, m := range markers {
+		if strings.Contains(s, m) {
+			return
+		}
+	}
+
+	// Some very short comment that can be skipped.
+	// Usually triggering on these results in false positive.
+	// Unless there is a very popular call like print/println.
+	cond := len(s) < len("quite too short") &&
+		!strings.Contains(s, "print") &&
+		!strings.Contains(s, "fmt.") &&
+		!strings.Contains(s, "log.")
+	if cond {
+		return
+	}
+
+	stmt := strparse.Stmt(s)
+	if stmt == strparse.BadStmt {
+		return // Most likely not a code
+	}
+
+	if !c.isPermittedStmt(stmt) {
+		c.warn(cg)
+	}
+}
+
+func (c *commentedOutCodeChecker) isPermittedStmt(stmt ast.Stmt) bool {
+	switch stmt := stmt.(type) {
+	case *ast.ExprStmt:
+		return c.isPermittedExpr(stmt.X)
+	case *ast.LabeledStmt:
+		return c.isPermittedStmt(stmt.Stmt)
+	case *ast.DeclStmt:
+		decl := stmt.Decl.(*ast.GenDecl)
+		return decl.Tok == token.TYPE
+	default:
+		return false
+	}
+}
+
+func (c *commentedOutCodeChecker) isPermittedExpr(x ast.Expr) bool {
+	// Permit anything except expressions that can be used
+	// with complete result discarding.
+	switch x := x.(type) {
+	case *ast.CallExpr:
+		return false
+	case *ast.UnaryExpr:
+		// "<-" channel receive is not permitted.
```

Comment:
But why?

Reply:
Here is explanation (may include into the code as a comment):
We start scanning from statements. If there is `ExprStmt`, we need
to determine whether it looks like commented-out code or not.
To do that, we use Go compiler property that forbids things like binary expr without using it's results. So, `// a + b` does not cause any warnings because it's `*ast.BinaryExpr`, and it's a good thing, because if we remove `//` this code would not compile, you can't use `a + b` without assigning it to something => so it's not commented-out code but some kind of comment that just happens looks like Go code.
Not so with function calls and recv expressions like `<- ch`.
They can be in place of statement without assignment (for side effects).
This means they could represent commented-out code.

Reply:
That's cool. It would be good to extend comment, but with a shorter addition, like: <- is permitted, 'cause `<-ch` can be an independent expression.

Reply:
Few lines above:
```go
	// Permit anything except expressions that can be used
	// with complete result discarding.
```

Reply:
kk